### DISCUSSION TITLE
Adding new integrity only Cipher to TLS1.3

### DISCUSSION THREAD

**Discussion by rajeev-0** [openssl/openssl#22902] [Q&A]
Hi, 
I am new to ciphers implementation and I am currently working on implementing an integrity-only cipher for TLS 1.3 (RFC 9150).
I noticed that the cipher code seems to be duplicated in the 'evp' module, for example, in files like 'e_chacha20_poly1305.c' and 'cipher_chacha20_poly1305.c'. I assume this is due to legacy support. 
Now, I am adding new integrity-only ciphers (TLS_SHA256_SHA256 and TLS_SHA384_SHA384) that utilize HMAC as a base. My question is, do I need to add code to the 'evp' module for these new ciphers as well? 

Thanks

**Comment by mattcaswell** [comment]
> I assume this is due to legacy support.

Correct. These are legacy EVP layer implementations.

> do I need to add code to the 'evp' module for these new ciphers as well?

No

> Now, I am adding new integrity-only ciphers (TLS_SHA256_SHA256 and TLS_SHA384_SHA384) that utilize HMAC as a base.

From a brief scan of RFC9150 this seems to be using existing digests and MACs. So I wouldn't expect to see new ciphers implemented in libcrypto/provider implementations at all. Plausibly there could be some modifications to existing algs to support some TLS level params. Mostly though I would expect these ciphersuites to require modifications in libssl.

**Reply by rajeev-0** [reply]
Thanks for fast response. 
Is there any way to populate the `nid` field of the `EVP_CIPHER` struct apart from using EVP_add_cipher()?

**Reply by mattcaswell** [reply]
I don't see why you would need to create a new EVP_CIPHER at all since (as I understand it) there is no encrypt/decrypt to do here. If you do need one you should implement it in the default provider which does not require you to have a nid and doe not need you to call EVP_add_cipher().

It's possible that you might need to make use of (and possibly modify in some way) the existing NULL cipher:

https://github.com/openssl/openssl/blob/master/providers/implementations/ciphers/cipher_null.c

**Reply by mattcaswell** [reply]
In fact thinking about it a bit more, adding AEAD support to the NULL cipher seems like a reasonable way to approach this.

**Reply by rajeev-0** [reply]
Yes, I took the the inspiration from `cipher_null.c`.[#22903](https://github.com/openssl/openssl/pull/22903)

But I struggle with setting the `nid` field of the `EVP_CIPHER` struct , which I needed to decide the `taglen` in  `ssl/tls13_enc.c:392` https://github.com/openssl/openssl/pull/22903/files#diff-6b861e097838cb9ed2fee42d13aa5a107edc54e1adcdfc1a42d2b24477c3a9dc.
I couldn't find any other way of doing it, apart from using EVP_add_cipher().
Is `nid` also part of legacy?

**Comment by DDvO** [comment]
This thread is no more relevant since the implementation in #22903 is finished.
